File: libs/power-log-parser/src/lib/parsers/removed-from-history-parser.ts
import { ActionParser } from '../action-parser';
import { MetaDataType } from '../enums';
import { GameEventProvider, GameEventHelper } from '../game-event';
import { MetaData, Node } from '../models';
import { GameState } from '../state/game-state';
import { ParserState, StateType } from '../state/parser-state';
import type { StateFacade } from '../state/state-facade';

export class RemovedFromHistoryParser implements ActionParser {
	readonly ParserName = 'RemovedFromHistoryParser';

	private GameState: GameState;
	private StateFacade: StateFacade;

	constructor(parserState: ParserState, stateFacade: StateFacade) {
		this.GameState = parserState.GameState;
		this.StateFacade = stateFacade;
	}

	AppliesOnNewNode(_node: Node, _stateType: StateType): boolean {
		return false;
	}

	AppliesOnCloseNode(node: Node, stateType: StateType): boolean {
		return (
			stateType === StateType.PowerTaskList &&
			node.Type === MetaData &&
			(node.Object as MetaData).Meta === (MetaDataType.HISTORY_REMOVE_ENTITIES as number)
		);
	}

	CreateGameEventProviderFromNew(_node: Node): GameEventProvider[] | null {
		return null;
	}

	CreateGameEventProviderFromClose(node: Node): GameEventProvider[] | null {
		const meta = node.Object as MetaData;
		const infos = meta.MetaInfo;
		const result: GameEventProvider[] = [];
		for (const info of infos) {
			const entity = this.GameState.CurrentEntities.get(info.Entity);
			if (entity?.Entity != null) {
				result.push(
					GameEventProvider.Create(
						meta.TimeStamp,
						'REMOVE_FROM_HISTORY',
						GameEventHelper.CreateProvider(
							'REMOVE_FROM_HISTORY',
							entity.CardId,
							entity.GetController(),
							entity.Id,
							this.StateFacade,
							null,
						),
						true,
						node,
					),
				);
			}
		}
		return result;
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Parsers/RemovedFromHistoryParser.cs
using HearthstoneReplays.Parser;
using HearthstoneReplays.Parser.ReplayData.Meta;
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.Entities;
using System.Collections.Generic;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using System;
using System.Linq;

namespace HearthstoneReplays.Events.Parsers
{
    public class RemovedFromHistoryParser : ActionParser
    {
        private GameState GameState { get; set; }
        private ParserState ParserState { get; set; }
        private StateFacade StateFacade { get; set; }

        public RemovedFromHistoryParser(ParserState ParserState, StateFacade facade)
        {
            this.ParserState = ParserState;
            this.GameState = ParserState.GameState;
            this.StateFacade = facade;
        }

        public bool AppliesOnNewNode(Node node, StateType stateType)
        {
            return false;
        }

        public bool AppliesOnCloseNode(Node node, StateType stateType)
        {
            return stateType == StateType.PowerTaskList
                && node.Type == typeof(MetaData)
                && (node.Object as MetaData).Meta == (int)MetaDataType.HISTORY_REMOVE_ENTITIES;
        }

        public List<GameEventProvider> CreateGameEventProviderFromNew(Node node)
        {
            return null;
        }

        public List<GameEventProvider> CreateGameEventProviderFromClose(Node node)
        {
            var meta = node.Object as MetaData;
            var infos = meta.MetaInfo;
            var result = new List<GameEventProvider>();
            foreach (var info in infos)
            {
                var entity = GameState.CurrentEntities.GetValueOrDefault(info.Entity);
                if (entity?.Entity != null)
                {
                    result.Add(GameEventProvider.Create(
                        meta.TimeStamp,
                        "REMOVE_FROM_HISTORY",
                        GameEvent.CreateProvider(
                            "REMOVE_FROM_HISTORY",
                            entity.CardId,
                            entity.GetController(),
                            entity.Id,
                            StateFacade,
                            null),
                        true,
                        node));
                }
            }
            return result;
        }
    }
}
